tokio/src/task/local.rs

```diff
@@ -686,7 +728,7 @@ impl<T: Future> Future for RunUntil<'_, T> {
 impl Shared {
     /// Schedule the provided task on the scheduler.
     fn schedule(&self, task: task::Notified<Arc<Self>>) {
-        CURRENT.with(|maybe_cx| match maybe_cx {
+        CURRENT.with(|maybe_cx| match maybe_cx.borrow().as_ref() {
             Some(cx) if cx.shared.ptr_eq(self) => {
                 cx.queue.push_back(task);
```

Comment:
Hmm, I may have changed my opinion on whether `Cell` is a good idea. Regardless, I am a bit worried about having us borrow the `RefCell` for too long. For example, in this case, the `RefCell` is borrowed until after the entire `match`, so the arbitrary user code inside the call to `self.waker.wake()` will run before the `RefCell` is released. Thus, if waking the waker involves calling `LocalSet::enter`, then that will result in a panic.

I would recommend that you define a function like this:
```
use std::cell::RefCell;
use std::rc::Rc;

fn clone_rc<T>(rc: &RefCell<Option<Rc<T>>>) -> Option<Rc<T>> {
    rc.borrow().clone()
}
```
And use it when matching on the context.

Reply:
can you elaborate why Cell is not a good idea, isn't it safer and will not lead to the issue as you mentioned above?

Reply:
Cell could be fine. It's just somewhat annoying to use it, though I guess you could define a similar `clone_rc` method:
```
fn clone_rc<T>(rc: &Cell<Option<Rc<T>>>) -> Option<Rc<T>> {
    let value = rc.take();
    let cloned = value.clone();
    rc.set(value);
    cloned
}
```

Reply:
update to use `Cell`, also rebase the latest changes from master